### PR TITLE
Create reusable package to manage manifests for use by CVO and oc

### PR DESCRIPTION
**Pull request by jottofar**
Create reusable package for use by CVO and 'oc adm release mirror'

### CODE REVIEW

File: pkg/manifest/manifest.go [hunk 1]
```diff
@@ -0,0 +1,118 @@
...
+	OriginalFilename string
+
+	Raw []byte
+	GVK schema.GroupVersionKind
+
+	Obj *unstructured.Unstructured
+}
+
+// UnmarshalJSON unmarshals bytes of single kubernetes object to Manifest.
+func (m *Manifest) UnmarshalJSON(in []byte) error {
```
Comment (sttts) [inline pkg/manifest/manifest.go]: why is this no constructor?

File: pkg/manifest/manifest.go [hunk 2]
```diff
@@ -0,0 +1,118 @@
...
+	"k8s.io/apimachinery/pkg/apis/meta/v1/unstructured"
+	"k8s.io/apimachinery/pkg/runtime/schema"
+	utilerrors "k8s.io/apimachinery/pkg/util/errors"
+	"k8s.io/apimachinery/pkg/util/yaml"
+	"k8s.io/client-go/kubernetes/scheme"
+)
+
+// Manifest stores Kubernetes object in Raw from a file.
+// It stores the GroupVersionKind for the manifest.
+type Manifest struct {
```
Comment (sttts) [inline pkg/manifest/manifest.go]: what are the invariants? Raw and Obj are always in sync and set?

File: pkg/manifest/manifest.go [hunk 3]
```diff
@@ -0,0 +1,118 @@
...
+	// OriginalFilename is set to the filename this manifest was loaded from.
+	// It is not guaranteed to be set or be unique, but we will set it when
+	// loading from disk to provide better debuggability.
+	OriginalFilename string
+
+	Raw []byte
+	GVK schema.GroupVersionKind
+
+	Obj *unstructured.Unstructured
```
Comment (sttts) [inline pkg/manifest/manifest.go]: Object

File: pkg/manifest/manifest.go [hunk 4]
```diff
@@ -0,0 +1,118 @@
...
+// Manifest stores Kubernetes object in Raw from a file.
+// It stores the GroupVersionKind for the manifest.
+type Manifest struct {
+	// OriginalFilename is set to the filename this manifest was loaded from.
+	// It is not guaranteed to be set or be unique, but we will set it when
+	// loading from disk to provide better debuggability.
+	OriginalFilename string
+
+	Raw []byte
+	GVK schema.GroupVersionKind
```
Comment (sttts) [inline pkg/manifest/manifest.go]: Obj gives the same information. Why a copy?

File: pkg/manifest/manifest.go [hunk 5]
```diff
@@ -0,0 +1,118 @@
...
+		return fmt.Errorf("expected manifest to decode into *unstructured.Unstructured, got %T", ud)
+	}
+
+	m.GVK = ud.GroupVersionKind()
+	m.Obj = ud.DeepCopy()
+	return nil
+}
+
+// Object returns underlying metav1.Object
+func (m *Manifest) Object() metav1.Object { return m.Obj }
```
Comment (sttts) [inline pkg/manifest/manifest.go]: with this, why a public Obj?

File: pkg/manifest/manifest.go (hunk 1 above)
Comment (jottofar) [reply pkg/manifest/manifest.go]: Not sure as this is existing code I simply migrated "as is" from CVO. The guidance I was given is to effect as little change as possible to the CVO code. I can add a constructor but forcing it's use by making the Manifest struct private would have big impact to CVO. Would providing a constructor but leaving Manifest public suffice?

File: pkg/manifest/manifest.go (hunk 2 above)
Comment (jottofar) [reply pkg/manifest/manifest.go]: They are always in sync with it's current use by CVO and my planned use by oc. The struct only gets set here by this call to ParseManifests in CVO:

https://github.com/jottofar/cluster-version-operator/blob/732ae625e991bf1447370b537caf9bcc22c850dc/pkg/payload/payload.go#L147

Also, both manifest.ManifestsFromFiles and manifest.UnmarshalJSON set both

File: pkg/manifest/manifest.go (hunk 3 above)
Comment (jottofar) [reply pkg/manifest/manifest.go]: Assume you want me to change Obj to Object? If so, is this absolutely required? Again, going back to goal of trying to minimize changes to CVO.

File: pkg/manifest/manifest.go (hunk 4 above)
Comment (jottofar) [reply pkg/manifest/manifest.go]: It is currently accessed by CVO through both Obj and Raw.

File: pkg/manifest/manifest.go (hunk 5 above)
Comment (jottofar) [reply pkg/manifest/manifest.go]: CVO accesses it directly in some cases:

https://github.com/jottofar/cluster-version-operator/blob/732ae625e991bf1447370b537caf9bcc22c850dc/pkg/cvo/cvo.go#L304

And in others Object() is used:

https://github.com/jottofar/cluster-version-operator/blob/732ae625e991bf1447370b537caf9bcc22c850dc/pkg/cvo/sync_worker.go#L918

File: pkg/manifest/manifest.go (hunk 2 above)
Comment (sttts) [reply pkg/manifest/manifest.go]: please document that. It's central for the type.

File: pkg/manifest/manifest.go (hunk 5 above)
Comment (sttts) [reply pkg/manifest/manifest.go]: change CVO

File: pkg/manifest/manifest.go (hunk 1 above)
Comment (jottofar) [reply pkg/manifest/manifest.go]: added a ctor

File: pkg/manifest/manifest.go (hunk 2 above)
Comment (jottofar) [reply pkg/manifest/manifest.go]: documented

File: pkg/manifest/manifest.go (hunk 4 above)
Comment (jottofar) [reply pkg/manifest/manifest.go]: documented structure

File: pkg/manifest/manifest.go (hunk 5 above)
Comment (jottofar) [reply pkg/manifest/manifest.go]: I removed the Object func

File: pkg/manifest/manifest.go (hunk 1 above)
Comment (sttts) [reply pkg/manifest/manifest.go]: why not make this func private?

File: pkg/manifest/manifest.go [hunk 6]
```diff
@@ -0,0 +1,132 @@
...
+	udi, _, err := scheme.Codecs.UniversalDecoder().Decode(in, nil, &unstructured.Unstructured{})
+	if err != nil {
+		return errors.Wrapf(err, "unable to decode manifest")
+	}
+	ud, ok := udi.(*unstructured.Unstructured)
+	if !ok {
+		return fmt.Errorf("expected manifest to decode into *unstructured.Unstructured, got %T", ud)
+	}
+
+	m.GVK = ud.GroupVersionKind()
```
Comment (sttts) [inline pkg/manifest/manifest.go]: here you see how senseless it is to store GVK independently.

File: pkg/manifest/manifest.go [hunk 7]
```diff
@@ -0,0 +1,132 @@
...
+	m.Obj = ud.DeepCopy()
```
Comment (sttts) [inline pkg/manifest/manifest.go]: why deepcopy?

File: pkg/manifest/manifest.go (hunk 1 above)
Comment (sttts) [reply pkg/manifest/manifest.go]: The answer to my questions: `UnmarshalJSON` is called by the yaml or json decoder. In other words, `Manifest` knows how to unmarshal itself. This means we must have this signature. And it also means the constructor is not used at all. Sorry for pointing you into the wrong direction.

Comment (jottofar) [reply pkg/manifest/manifest.go]: Removed ctor and documented UnmarshalJSON.

File: pkg/manifest/manifest.go (hunk 7 above)
Comment (jottofar) [reply pkg/manifest/manifest.go]: Must be required. I removed the DeepCopy and existing code gets a SIGSEGV nil pointer when trying to access Obj:

`src := fmt.Sprintf("the config map %s/%s", ms[i].Obj.GetNamespace(), ms[i].Obj.GetName())`

Comment (sttts) [reply pkg/manifest/manifest.go]: That's odd. Of course I meant `m.Obj = ud`. Doesn't work?

Comment (jottofar) [reply pkg/manifest/manifest.go]: Yes, when I commented out line 65 'm.Obj = ud.DeepCopy()' it failed with SIGSEGV.

Comment (jottofar) [reply pkg/manifest/manifest.go]: Changed and does work

**Pull request merged**